src/mpid/ch4/shm/xpmem/xpmem_send.h

```diff
@@ -28,18 +29,31 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_lmt_isend(const void *buf, MPI_Aint cou
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_LMT_ISEND);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_LMT_ISEND);
 
-    sreq = MPIDIG_request_create(MPIR_REQUEST_KIND__SEND, 2);
-    MPIR_ERR_CHKANDSTMT((sreq) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
-    *request = sreq;
-
     MPIDI_Datatype_check_contig_size_lb(datatype, count, is_contig, data_sz, true_lb);
 
+    if (packet_type == MPIDI_SHM_XPMEM_SEND_LMT_RTS) {
+        sreq = MPIDIG_request_create(MPIR_REQUEST_KIND__SEND, 2);
+        MPIR_ERR_CHKANDSTMT((sreq) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
+        *request = sreq;
+
+        slmt_req_hdr->sreq_ptr = (uint64_t) sreq;
+        MPIDIG_REQUEST(sreq, buffer) = (void *) buf;
+        MPIDIG_REQUEST(sreq, datatype) = datatype;
+        MPIDIG_REQUEST(sreq, count) = count;
+        MPIDI_XPMEM_REQUEST(sreq, counter) = NULL;
+    } else {
+        /* Receiver replies CTS packet */
+        slmt_req_hdr->sreq_ptr = MPIDI_XPMEM_REQUEST(*request, sreq_ptr);
+        slmt_req_hdr->rreq_ptr = (uint64_t) * request;
+    }
```

Comment:
`MPIDI_XPMEM_lmt_coop_isend` should only be called by sender. If you want to reuse the header setup code (e.g., XPMEM internal info), you can define a common subroutine for it.

Moreover, I do not think the CTS packet needs `message matching info`. You define use a separate packet header for it.